vmm/src/device_manager.rs

```diff
@@ -3552,11 +3570,28 @@ impl DeviceManager {
             .pci_device_handle
             .ok_or(DeviceManagerError::MissingPciDevice)?;
         let (pci_device, bus_device, virtio_device) = match pci_device_handle {
-            PciDeviceHandle::Vfio(vfio_pci_device) => (
-                Arc::clone(&vfio_pci_device) as Arc<Mutex<dyn PciDevice>>,
-                Arc::clone(&vfio_pci_device) as Arc<Mutex<dyn BusDevice>>,
-                None as Option<VirtioDeviceArc>,
-            ),
+            PciDeviceHandle::Vfio(vfio_pci_device) => {
+                {
+                    // Unregister the VFIO mapping handler from all virtio-mem
+                    // devices.
+                    let dev = vfio_pci_device.lock().unwrap();
+                    if !dev.iommu_attached() {
+                        for virtio_mem_device in self.virtio_mem_devices.iter() {
+                            virtio_mem_device
+                                .lock()
+                                .unwrap()
+                                .remove_dma_mapping_handler(pci_device_bdf)
+                                .map_err(DeviceManagerError::RemoveDmaMappingHandlerVirtioMem)?;
+                        }
+                    }
+                }
```

Comment:
@sboeuf Did I need to restore this part? Or is this handled some other way?

Reply:
Well for this part, you're gonna have to handle it for vfio-user the way we used to handle it for VFIO. And for VFIO I was thinking we could restore the previous code, but we would need to do some reference counting internally from virtio-mem.
From a VFIO perspective, we can't afford to remove the handler if we simply remove one VFIO device. Instead we must know how many device share the same handler and count. This way we don't invoke multiple times the same handler, but on the other hand we don't remove a handler that might still be needed by other devices.

Reply:
> Well for this part, you're gonna have to handle it for vfio-user the way we used to handle it for VFIO. 

Yeh. I do that below.

>And for VFIO I was thinking we could restore the previous code, but we would need to do some reference counting internally from virtio-mem. From a VFIO perspective, we can't afford to remove the handler if we simply remove one VFIO device. Instead we must know how many device share the same handler and count. This way we don't invoke multiple times the same handler, but on the other hand we don't remove a handler that might still be needed by other devices.

The handler is already in an `Arc` (which is already ref counted). I think if we create one `VfioDmaMapping` per container (when the container is first created) and then add it for every device to the list of mappings in virtio-mem then that would work. The only issue is we'd need to ignore subsequent requests from virtio-mem to map the same address (which we could do in the VfioDmaMapping.).

Reply:
AFAICT at most there is only ever one container added to the virtio-mem mapping (if there are other containers then they are going through the virtio-iommu so don't get added to virtio-mem) and that container is never destroyed (it it saved on `DeviceManager`. And since the mapping is based on the container we should never need to remove the mapping in the Vfio case (i.e. the code above isn't needed.)

For clarity I think modifying `VirtioMemDevice::AddDmaMappingHandler` to take an `enum { Container, Device(u32) }` and storing that would be neat.

Reply:
That would be neat indeed, I'm just worried that if we provide `Container` for the VFIO case, and if we unplug all VFIO devices associated with the container, we might end up in a case where the container might be dropped, hence not being valid anymore to receive mapping updates from virtio-mem devices. It's a rather tricky case but something to keep in mind.

Reply:
Ok thanks for checking about the lifetime. I'm wondering from a VFIO perspective if the VFIO container will accept map/unmap updates if no devices are attached to it. Nothing to do with Rust, but with the VFIO driver itself.

Reply:
I think it should be OK (there is an explicit check in the kernel) i'm sure that the changes I propose actually change this behaviour relative to what was there before.